tests/test_kubernetes.py

```diff
@@ -0,0 +1,166 @@
+def runfile(request):
+    if request.param == 'gcloud':
+        return 'gs://my-bucket/runfile.py'
+    if request.param == 'local':
+        return '/inside/docker/dir/runfile.py'
+
+
+@pytest.fixture
+def jobname():
+    return 'my-job-name'
+
+
+@pytest.fixture(params=['my-project', None])
+def google_cloud_project(request):
+    return request.param
+
+
+@pytest.fixture(params=["default", "manual_limit"])
+def memory_args(request):
+    return {
+        "default": MemoryArgs(3.6, None, 3.6),
+        "manual_limit": MemoryArgs(3.6, 4.0, 4.0),
+    }[request.param]
+
+
+@pytest.fixture(params=["1_cpu", "2.5_cpu"])
+def cpu_count(request):
+    if request.param == "1_cpu":
+        return 1
+    elif request.param == "2.5_cpu":
+        return 2.5
+
+
+@pytest.fixture(params=['my-secret-name', None])
+def gcp_secret(request):
+    return request.param
+
+
+@pytest.fixture(
+    params=['Always', 'IfNotPresent']
+)
+def image_pull_policy(request):
+    return request.param
+
+
+def test_local_config_rejected(outdir, docker_image, runfile):
+    with pytest.raises(ValueError):
+        fv3config.fv3run._kubernetes._get_job(
+            '/local/dir/config.yml', outdir, docker_image, runfile)
+
+
+def test_local_outdir_rejected(config_location, docker_image, runfile):
+    with pytest.raises(ValueError):
+        fv3config.fv3run._kubernetes._get_job(
+            config_location, '/local/outdir', docker_image, runfile)
+
+
+def test_get_job(
+        config_location, outdir, docker_image, runfile, jobname,
+        google_cloud_project, memory_args, cpu_count, gcp_secret, image_pull_policy):
+    job = fv3config.fv3run._kubernetes._get_job(
+        config_location, outdir, docker_image, runfile, jobname,
+        google_cloud_project, memory_args.memory_gb, memory_args.memory_gb_limit_in,
+        cpu_count, gcp_secret, image_pull_policy)
+    _check_job(job, jobname)
+    job_spec = job.spec
+    _check_job_spec(job_spec)
+    pod_spec = job_spec.template.spec
+    assert pod_spec.restart_policy == 'Never'
+    container = pod_spec.containers[0]
+    _check_command(container.command, outdir, config_location, runfile)
+    assert container.image == docker_image
+    assert container.image_pull_policy == image_pull_policy
+    _check_secret_mount(gcp_secret, container, pod_spec)
+    if gcp_secret is not None:
+        _check_env(container.env, 'GOOGLE_APPLICATION_CREDENTIALS', '/secret/gcp-credentials/key.json')
+        _check_env(container.env, 'CLOUDSDK_AUTH_CREDENTIAL_FILE_OVERRIDE', '/secret/gcp-credentials/key.json')
+    else:
+        _check_env(container.env, 'GOOGLE_APPLICATION_CREDENTIALS', None)
+        _check_env(container.env, 'CLOUDSDK_AUTH_CREDENTIAL_FILE_OVERRIDE', None)
+    _check_env(container.env, 'GOOGLE_CLOUD_PROJECT', google_cloud_project)
+    _check_resource_requirements(
+        container, memory_args.memory_gb, memory_args.memory_gb_limit, cpu_count)
+
+
+def _check_job(job, jobname):
```

Comment:
I think you should change "check" to "assert" in all these functions. In this case, "check" and "assert" have the same meaning, so it is confusing that we use `assert` in some places, and `check` in others. *Clean code* makes the point that the same concept shouldn't have two names.

Reply:
They are distinct concepts. You "check" a noun while you "assert" a statement. "assert job" implies asserting bool(job), in other words that job exists. I think I have been consistent in using "assert_{expression}" everywhere else, "check_{object}" has generally not been necessary because we've avoided writing many objects.

I really dislike having a "check_{object}" test at all because it implies we're testing multiple expressions about the object and I'd rather have one test for each, but there are way too many super simple tests for that to be good here.

It's also generally bad to "test that the object you instantiated is what you supplied as arguments", because it generally implies the object you instantiated should be simpler. It's essential here because we're wrapping some pretty gnarly google API we aren't allowed to change into a much simpler job creation API.

I could have used "_assert_job_is_correct", but I felt "_check_job" was a more succinct way of saying this. Would you prefer I change the structure of the names to `_assert_{noun}_is_correct`?

Reply:
Okay. That is fine. I agree that it's hard to cleanly test a deeply nested structure like the google api. So this is more of an integration test anyway.